File: src/vector-index.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import type { SessionRef } from "./types.js";
import { ensureModel, embedText } from "./embed.js";

const INDEX_DIR = join(homedir(), ".agent-hop");
const INDEX_PATH = join(INDEX_DIR, "index.json");
const CHUNK_CHARS = 2000; // MiniLM works best on focused text, not huge blobs
const CHUNK_OVERLAP = 200; // avoid splitting a relevant sentence exactly at a chunk boundary

interface ChunkEntry {
  vector: number[];
}

interface SessionEntry {
  key: string; // `${tool}:${sessionId}`
  sourceMtime: number;
  chunks: ChunkEntry[];
}

function loadIndex(): Map<string, SessionEntry> {
  if (!existsSync(INDEX_PATH)) return new Map();
  try {
    const entries: SessionEntry[] = JSON.parse(readFileSync(INDEX_PATH, "utf-8"));
    return new Map(entries.map((e) => [e.key, e]));
  } catch {
    return new Map();
  }
}

function saveIndex(index: Map<string, SessionEntry>): void {
  mkdirSync(INDEX_DIR, { recursive: true });
  writeFileSync(INDEX_PATH, JSON.stringify([...index.values()]));
}

export function chunkSession(session: SessionRef): string[] {
  const body = session.body ?? session.snippet;
  const full = `${session.title}\n${body}`;
  if (full.length <= CHUNK_CHARS) return [full];

  const chunks: string[] = [];
  let start = 0;
  while (start < full.length) {
    chunks.push(full.slice(start, start + CHUNK_CHARS));
    start += CHUNK_CHARS - CHUNK_OVERLAP;
  }
  return chunks;
}

/**
 * Reads whatever's already been embedded from disk -- no new embedding work,
 * no blocking. Sessions that haven't been indexed yet simply don't get a
 * score (caller treats that as "no semantic signal yet, BM25 still applies").
 * This is what live search calls, so a search is never gated on indexing.
 */
export function getCachedSemanticScores(
  sessions: SessionRef[],
  queryVec: Float32Array,
  cosineSimilarity: (a: Float32Array, b: Float32Array) => number
): Map<string, number> {
  const index = loadIndex();
  const scores = new Map<string, number>();
  for (const s of sessions) {
    const key = `${s.tool}:${s.sessionId}`;
    const entry = index.get(key);
    if (!entry) continue;
    let best = -1;
    for (const chunk of entry.chunks) {
      const sim = cosineSimilarity(queryVec, Float32Array.from(chunk.vector));
      if (sim > best) best = sim;
    }
    scores.set(key, best);
  }
  return scores;
}

// A session that's still being actively written to (e.g. the very
// conversation you're having with an agent right now) has its mtime change
// on nearly every turn. Without this, every single search would see it as
// "changed since last index" and kick off a fresh background embedding run
// -- which then goes stale again within seconds, so the next search kicks
// off *another* one, forever, for as long as you keep using that agent.
// Skipping anything modified in the last few minutes means we wait for it
// to settle instead of chasing a moving target.
const SETTLE_WINDOW_MS = 3 * 60 * 1000;

function needsEmbedding(s: SessionRef, index: Map<string, SessionEntry>): boolean {
  if (Date.now() - s.updatedAt < SETTLE_WINDOW_MS) return false;
  const existing = index.get(`${s.tool}:${s.sessionId}`);
  return !existing || existing.sourceMtime !== s.updatedAt;
}

/** True if any session needs (re-)embedding since the last index build. */
export function hasPendingWork(sessions: SessionRef[]): boolean {
  const index = loadIndex();
  return sessions.some((s) => needsEmbedding(s, index));
}

const LOCK_PATH = join(INDEX_DIR, "indexing.lock");

function isLockStale(): boolean {
  try {
    const pid = Number(readFileSync(LOCK_PATH, "utf-8").trim());
    process.kill(pid, 0); // throws if the process doesn't exist
    return false;
  } catch {
    return true; // no lock, unparseable, or process is gone
  }
}

/**
 * Actually does the embedding work -- sequential, in-process. Meant to run
 * inside the detached background process (see background-index.ts), never
 * inline in an interactive search. Writes progressively so a killed/crashed
 * run doesn't lose everything already completed.
 */
export async function buildIndex(sessions: SessionRef[], onProgress?: (done: number, total: number) => void): Promise<void> {
  mkdirSync(INDEX_DIR, { recursive: true });
  if (existsSync(LOCK_PATH) && !isLockStale()) return; // another run already in progress
  writeFileSync(LOCK_PATH, String(process.pid));

  try {
    const index = loadIndex();
    const toEmbed = sessions.filter((s) => needsEmbedding(s, index));
    if (toEmbed.length === 0) return;

    await ensureModel();
    for (let i = 0; i < toEmbed.length; i++) {
      const s = toEmbed[i];
      const key = `${s.tool}:${s.sessionId}`;
      const texts = chunkSession(s);
      const chunks: ChunkEntry[] = [];
      for (const text of texts) {
        const vec = await embedText(text);
        chunks.push({ vector: Array.from(vec) });
      }
      index.set(key, { key, sourceMtime: s.updatedAt, chunks });
      onProgress?.(i + 1, toEmbed.length);
      // save incrementally every few sessions so progress isn't lost if
      // this background process gets killed partway through
      if (i % 10 === 0) saveIndex(index);
    }

    const liveKeys = new Set(sessions.map((s) => `${s.tool}:${s.sessionId}`));
    for (const key of [...index.keys()]) {
      if (!liveKeys.has(key)) index.delete(key);
    }
    saveIndex(index);
  } finally {
    try {
      unlinkSync(LOCK_PATH);
    } catch {
      // already gone, fine
    }
  }
}

File: src/vector_index.rs
//! Port of src/vector-index.ts -- the persistent semantic-search index:
//! chunking, cached-score lookup (never blocks search), and the actual
//! (re-)embedding work meant to run in a detached background process.

use crate::adapters::SessionRef;
use crate::embed;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

fn index_dir() -> PathBuf {
    dirs::home_dir().unwrap_or_default().join(".agent-hop")
}
fn index_path() -> PathBuf {
    index_dir().join("index.json")
}
fn lock_path() -> PathBuf {
    index_dir().join("indexing.lock")
}

const CHUNK_CHARS: usize = 2000; // MiniLM works best on focused text, not huge blobs
const CHUNK_OVERLAP: usize = 200; // avoid splitting a relevant sentence exactly at a chunk boundary

#[derive(Serialize, Deserialize, Clone)]
struct ChunkEntry {
    vector: Vec<f32>,
}

#[derive(Serialize, Deserialize, Clone)]
struct SessionEntry {
    key: String, // "{tool}:{sessionId}"
    #[serde(rename = "sourceMtime")]
    source_mtime: i64,
    chunks: Vec<ChunkEntry>,
}

fn session_key(s: &SessionRef) -> String {
    format!("{}:{}", s.tool.slug(), s.session_id)
}

fn load_index() -> HashMap<String, SessionEntry> {
    let Ok(text) = std::fs::read_to_string(index_path()) else { return HashMap::new() };
    let Ok(entries) = serde_json::from_str::<Vec<SessionEntry>>(&text) else { return HashMap::new() };
    entries.into_iter().map(|e| (e.key.clone(), e)).collect()
}

fn save_index(index: &HashMap<String, SessionEntry>) -> anyhow::Result<()> {
    std::fs::create_dir_all(index_dir())?;
    let entries: Vec<&SessionEntry> = index.values().collect();
    std::fs::write(index_path(), serde_json::to_string(&entries)?)?;
    Ok(())
}

fn char_len(s: &str) -> usize {
    s.chars().count()
}

pub fn chunk_session(session: &SessionRef) -> Vec<String> {
    let body = session.body.clone().unwrap_or_else(|| session.snippet.clone());
    let full = format!("{}\n{}", session.title, body);
    if char_len(&full) <= CHUNK_CHARS {
        return vec![full];
    }
    let chars: Vec<char> = full.chars().collect();
    let mut chunks = Vec::new();
    let mut start = 0usize;
    while start < chars.len() {
        let end = (start + CHUNK_CHARS).min(chars.len());
        chunks.push(chars[start..end].iter().collect());
        start += CHUNK_CHARS - CHUNK_OVERLAP;
    }
    chunks
}

/// Reads whatever's already been embedded from disk -- no new embedding
/// work, no blocking. Sessions that haven't been indexed yet simply don't
/// get a score (caller treats that as "no semantic signal yet, BM25 still
/// applies"). This is what live search calls, so a search is never gated
/// on indexing.
pub fn get_cached_semantic_scores(sessions: &[SessionRef], query_vec: &[f32]) -> HashMap<String, f32> {
    let index = load_index();
    let mut scores = HashMap::new();
    for s in sessions {
        let key = session_key(s);
        let Some(entry) = index.get(&key) else { continue };
        let mut best = -1.0f32;
        for chunk in &entry.chunks {
            let sim = embed::cosine_similarity(query_vec, &chunk.vector);
            if sim > best {
                best = sim;
            }
        }
        scores.insert(key, best);
    }
    scores
}

// A session still being actively written to has its mtime change on
// nearly every turn. Without this, every search would see it as "changed
// since last index" and kick off a fresh background embedding run that
// goes stale again within seconds. Skipping anything modified in the last
// few minutes means we wait for it to settle instead of chasing a moving
// target.
const SETTLE_WINDOW_MS: i64 = 3 * 60 * 1000;

fn needs_embedding(s: &SessionRef, index: &HashMap<String, SessionEntry>) -> bool {
    let now_ms = chrono::Utc::now().timestamp_millis();
    if now_ms - s.updated_at < SETTLE_WINDOW_MS {
        return false;
    }
    match index.get(&session_key(s)) {
        Some(existing) => existing.source_mtime != s.updated_at,
        None => true,
    }
}

/// True if any session needs (re-)embedding since the last index build.
pub fn has_pending_work(sessions: &[SessionRef]) -> bool {
    let index = load_index();
    sessions.iter().any(|s| needs_embedding(s, &index))
}

fn is_lock_stale() -> bool {
    let Ok(text) = std::fs::read_to_string(lock_path()) else { return true };
    let Ok(pid) = text.trim().parse::<u32>() else { return true };
    process_exists(pid)
        .map(|exists| !exists)
        .unwrap_or(true)
}

#[cfg(unix)]
fn process_exists(pid: u32) -> anyhow::Result<bool> {
    // kill(pid, 0) -- signal 0 checks existence without actually signaling.
    let result = unsafe { libc::kill(pid as i32, 0) };
    Ok(result == 0)
}

#[cfg(not(unix))]
fn process_exists(_pid: u32) -> anyhow::Result<bool> {
    Ok(false)
}

/// Actually does the embedding work -- sequential, in-process. Meant to run
/// inside a detached background process, never inline in an interactive
/// search. Writes progressively so a killed/crashed run doesn't lose
/// everything already completed.
pub async fn build_index(sessions: &[SessionRef]) -> anyhow::Result<()> {
    std::fs::create_dir_all(index_dir())?;
    if lock_path().exists() && !is_lock_stale() {
        return Ok(()); // another run already in progress
    }
    std::fs::write(lock_path(), std::process::id().to_string())?;

    let result = (|| async {
        let mut index = load_index();
        let to_embed: Vec<&SessionRef> = sessions.iter().filter(|s| needs_embedding(s, &index)).collect();
        if to_embed.is_empty() {
            return Ok(());
        }

        embed::ensure_model(|_msg| {}).await?;
        for (i, s) in to_embed.iter().enumerate() {
            let key = session_key(s);
            let texts = chunk_session(s);
            let mut chunks = Vec::new();
            for text in &texts {
                let vec = embed::embed_text(text)?;
                chunks.push(ChunkEntry { vector: vec });
            }
            index.insert(key.clone(), SessionEntry { key, source_mtime: s.updated_at, chunks });
            // save incrementally every few sessions so progress isn't lost
            // if this background process gets killed partway through
            if i % 10 == 0 {
                save_index(&index)?;
            }
        }

        let live_keys: std::collections::HashSet<String> = sessions.iter().map(session_key).collect();
        index.retain(|k, _| live_keys.contains(k));
        save_index(&index)?;
        Ok::<(), anyhow::Error>(())
    })()
    .await;

    let _ = std::fs::remove_file(lock_path());
    result
}
